### ISSUE TITLE
Protect downstream clients from flaky websites and crashing spiders

### ISSUE THREAD

**Issue by brawer** [alltheplaces/alltheplaces#4351]
Currently, the weekly AllThePlaces dump contains the output from the last weekly run of each spider. However, spiders tend to be flaky. For example:

* the scraped website can temporarily be down;
* the website may be up, but they had a database problem, so the scraper produces a valid GeoJSON file with zero features;
* the spider crashes, leaving partial output files behind. This was pointed out by @iandees in a review comment on pull request https://github.com/alltheplaces/alltheplaces/issues/4339#issuecomment-1352777878.

It would be nice if downstream users didn’t have to worry about all this. Proposal: Instead of building the weekly dumpfile by calling `zip` on the raw weekly harvest, construct the zipfile from a small Python script. For each spider, this script would check if the current harvest is good (ie., the output has been completely written, and it contains at least one GeoJSON feature). If the current week’s harvest is good, the script would copy it into the final dump file. Otherwise, the script would try to find a valid output from previous runs of the same spider, and use that.

The script should probably impose a limit on how much it wants to go back in time; perhaps 3 months. However, it should go back in time only for scrapers that haven’t been deleted from the ATP codebase. For example, when spider `aldi_ch` gets replaced by `aldi_sud_ch` (as happened on 2022-10-22), the harvest from the deleted spider should _not_ get included into weekly dump files anymore.

**Comment by mjoe999** [comment]
_"weekly harvest"_ .... what a great term!

From my perspective no problem with it at all. We would still have the "truth":

https://www.alltheplaces.xyz/spiders.html

and you might also wish to keep this "honest":

https://www.alltheplaces.xyz/builds.html

But the "in your face" download on the "home page"  could be as you suggest and in fact where there is a big deviation in POI count between weeks (hello "western_union" spider) take the maximum within a period.

There's also the argument that as the number of spiders increases and I suspect we wish to be a bit less aggressive in our request rate in a good number of cases them the notion of a "weekly run" might start to blur ... but that would be a way away!

Bottom line is I'm also quite happy for things to remain "as is".

**Comment by iandees** [comment]
Yep, this is a good idea!

Joe, your point about picking which "harvest" to use is a good one. I like the default of picking the run with the maximum results in the time period, but worry that would hide stores that actually close over the course of time. It might be that we need to pick the latest complete run, even if it means we expose some of our flakiness?

Handling spider renames/deletions is interesting, too. We should probably _not_ include results from spiders that we delete, since it's likely that those stores don't exist anymore (e.g. Toys'r'Us).

For renaming spiders we should keep track of enough information to "redirect" old spiders to new. For example, maybe when we renamed "foe" to "foo" (woops misspelled!) to "foo_us" (oh, it's US-only) the new "foo_us" spider has a field like `deprecates = ['foe', 'foo']` so that we could programmatically connect them/

**Comment by Cj-Malone** [comment]
I wonder if keeping track of spiders being renamed is important enough to warrant the effort?

If we enforce the standard when merging, then we only have some legacy ones that need updating. It should normal out soon enough.

I could probably be convinced to do a spider name clean up, like I did to class names a while ago... I'm not entirely sure it's worth it, but I'd do it if we wanted it.

Plus we've got Wikidata, so how useful is spider name to consumers?

**Comment by Cj-Malone** [comment]
Related #1555

**Comment by iandees** [comment]
Yep, good point – it's probably overkill. I was mostly thinking that the deprecates field would help with picking the data to include, not just for data consumers.
